llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,398 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Object/SYCLBIN.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace llvm::object;
+
+namespace {
+
+class SYCLBINBlockReader {
+protected:
+  SYCLBINBlockReader(const char *Data, size_t Size) : Data{Data}, Size{Size} {}
+
+  Error ReadSizeCheck(size_t ByteOffset, size_t ReadSize) {
+    if (ByteOffset + ReadSize > Size)
+      return createStringError(inconvertibleErrorCode(),
+                               "Unexpected file contents size.");
+    return Error::success();
+  }
+
+  const char *Data = nullptr;
+  size_t Size = 0;
+};
+
+class SYCLBINHeaderBlockReader : public SYCLBINBlockReader {
+public:
+  SYCLBINHeaderBlockReader(const char *Data, size_t Size)
+      : SYCLBINBlockReader(Data, Size) {}
+
+  template <typename HeaderT>
+  Expected<const HeaderT *> GetHeaderPtr(size_t ByteOffset) {
+    if (Error ReadSizeError = ReadSizeCheck(ByteOffset, sizeof(HeaderT)))
+      return ReadSizeError;
+    return reinterpret_cast<const HeaderT *>(Data + ByteOffset);
+  }
+};
+
+class SYCLBINByteTableBlockReader : public SYCLBINBlockReader {
+public:
+  SYCLBINByteTableBlockReader(const char *Data, size_t Size)
+      : SYCLBINBlockReader(Data, Size) {}
+
+  Expected<StringRef> GetBinaryBlob(size_t ByteOffset, uint64_t BlobSize) {
+    if (Error ReadSizeError = ReadSizeCheck(ByteOffset, BlobSize))
+      return ReadSizeError;
+    return llvm::StringRef{Data + ByteOffset, BlobSize};
+  }
+
+  Expected<std::unique_ptr<llvm::util::PropertySetRegistry>>
+  GetMetadata(size_t ByteOffset, uint64_t MetadataSize) {
+    Expected<StringRef> BlobOrError = GetBinaryBlob(ByteOffset, MetadataSize);
+    if (!BlobOrError)
+      return BlobOrError.takeError();
+
+    std::string PropStr{BlobOrError->str()};
+    auto PropMemBuff = llvm::MemoryBuffer::getMemBuffer(PropStr);
+    auto ErrorOrProperties =
+        llvm::util::PropertySetRegistry::read(PropMemBuff.get());
+    if (!ErrorOrProperties)
+      return ErrorOrProperties.takeError();
+    return {std::move(*ErrorOrProperties)};
+  }
+};
+
+} // namespace
+
+Expected<std::optional<SYCLBIN::IRModuleHeaderType>>
+SYCLBIN::createIRModuleHeader(const SYCLBIN::ModuleDesc &Desc,
+                              const module_split::SplitModule &SM,
+                              SmallString<0> &MetadataByteTable,
+                              raw_svector_ostream &MetadataByteTableOS,
+                              SmallString<0> &BinaryByteTable,
+                              raw_svector_ostream &BinaryByteTableOS) {
+  // If no arch string is present, the module must be IR.
+  if (!Desc.ArchString.empty())
+    return std::nullopt;
```

Comment:
i see this answers my previous question about when we can have no error empty optional, but i dont understand how the comment matches the return value

Reply:
Good point! They were taken from another context where I think they made a little more sense. They've been reworded.